src/http_server/api/v1/health.c

```diff
@@ -245,6 +303,104 @@ static int cleanup_metrics()
     return c;
 }
 
+static int check_throughput_health(uint64_t in_records,
+                                   uint64_t out_records,
+                                   struct mk_list *sample_list,
+                                   int sample_count,
+                                   double out_in_ratio_threshold) {
+    struct flb_time tp;
+    uint64_t timestamp_seconds;
+    uint64_t in_rate;
+    uint64_t out_rate;
+    struct mk_list *tmp;
+    struct mk_list *head;
+    double out_in_ratio;
+    struct flb_hs_throughput_sample *entry;
+    struct flb_hs_throughput_sample *prev;
+    struct flb_hs_throughput_sample *sample;
+    struct flb_hs_throughput_sample *last_sample = NULL;
+    bool healthy;
+    bool rv;
+
+    flb_time_get(&tp);
+    timestamp_seconds = flb_time_to_nanosec(&tp) / 1000000000;
+
+    if (mk_list_is_empty(sample_list) != 0) {
+        last_sample = mk_list_entry_last(sample_list,
+                                         struct flb_hs_throughput_sample,
+                                         _head);
+    }
+
+    if (last_sample &&
+        in_records == last_sample->in_records &&
+        out_records == last_sample->out_records) {
+        flb_debug("[api/v1/health/throughput]: no changes since last check");
+        /* don't collect another sample unless either in_records or out_records have
+         * changed*/
+        goto check;
+    }
+
+    sample = flb_malloc(sizeof(struct flb_hs_throughput_sample));
```

Comment:
Could we please use an array for samples?

That way we could make one single `flb_calloc` call to allocate the entire sample list when initializing the plugin and then use a rolling index to overwrite the oldest record.

That would prevent fragmentation, minimize unnecessary allocations and make the overall code way simpler.

Reply:
This array + rolling index data structure is called a ring buffer. Is there an existing implementation in the codebase or should I implement one myself?

Reply:
Well, it's the same underlying concept without the fluff, I wouldn't abstract it or use the new ring buffer to keep it as lean as possible, just an array and 2 indexes.

Reply:
I will wait for this PR to be validated and approved before doing this. Doesn't make sense to optimize now if we don't even know if it will be merged.

Reply:
I ended up implementing your suggestion. Even though the code is more complex (due to the new ring buffer implementation), I'm sure it is now more efficient due to reduced allocations.